persist/adapter.go

```diff
@@ -65,3 +66,44 @@ type Adapter interface {
 	// This is part of the Auto-Save feature.
 	RemoveFilteredPolicy(sec string, ptype string, fieldIndex int, fieldValues ...string) error
 }
+
+type ContextAdapter interface {
+	Adapter
+
+	// LoadPolicy loads all policy rules from the storage.
+	LoadPolicyWithContext(ctx context.Context, model model.Model) error
+}
+
+type WrappedAdapter struct {
```

Comment:
Don't add Wrapadapter. Just add LoadPolicyWithContext interface like other adapter (BatchAdapter, FilterAdapter)

Reply:
This is also what I initially thought, but this would mean adapters need to support both methods instead of only legacy or context versions and we need to keep track of what the user called. So I wondered is it worth the effort?

Reply:
@johanneswuerbach upvote for @tangyang9464

Reply:
> This is also what I initially thought, but this would mean adapters need to support both methods instead of only legacy or context versions and we need to keep track of what the user called. So I wondered is it worth the effort?

@johanneswuerbach I think ``Context`` just offers an option and not a must. Wrapadapter is a burden to ``adapter``

Reply:
As outlined in https://github.com/casbin/casbin/issues/982, my biggest concern is significantly increasing the code size (+ tests) by having context and context-less methods.

To maintain backwards compatibility my proposal would be to always use context internally, so all context-less methods are essentially just wrappers around the context version, e.g.

```go
func (e *Enforcer) LoadPolicy() error {
  return e.LoadPolicyWithContext(context.Background())
}

func (e *Enforcer) LoadPolicyWithContext(ctx context.Context) error {
  // code
}
```

While this greatly simplifies the implementation (and requires almost zero new tests), the problem with this approach is that `LoadPolicy` would now also attempt to call the adapter `LoadPolicyWithContext`, so we either need to somehow remember the context-less first invocation (maybe using a [context value](https://pkg.go.dev/context#example-WithValue)) or drop the context again when the adapter isn't supporting context yet.

This was the idea behind the WrapAdapter, wrap adapters, which don't support context yet and simply drop the context. Having such a wrapper would also allow new adapters to only provide context methods, but casbin users could still call the legacy `LoadPolicy` if the desire.

Reply:
``LoadPolicyWithContext`` is called when the adapter supports ``context``, if not, calling ``LoadPolicy`` instead. This is almost the same as the WrapAdapter effect. I don't think it's necessary to wrap another layer with WrapAdapter. The new adapter needs to implement ``LoadPolicyWithContext``, and the implementation of ``LoadPolicy`` just calls ``LoadPolicyWithContext``. Of course, the new adapter can theoretically not implement ``LoadPolicy``, because the enforcer will not call it, but for compatibility, obviously we cannot delete the LoadPolicy interface. So it seems that the ``Wrapadapter`` has little effect

Reply:
Pushed an updated version :-)

Reply:
Any update here?